qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/QuteCodeActionForUnknownMethod.java

```diff
@@ -79,86 +80,62 @@ public void doCodeActions(CodeActionRequest request, List<CompletableFuture<Void
+		Set<String> existingProperties = new HashSet<>();
 
-		// Resolve methods defined in Java type
-		Collection<String> availableMethodNames = collectJavaTypeMethodNames(part, template, projectUri, baseResolvedType, filter, nativeImageSettings);
-		for (String methodName : availableMethodNames) {
-			availableValues.add(methodName);
-		}
+		// Collect similar code action for methods defined in Java type
+		collectSimilarCodeActionsForJavaMethods(part, template, projectUri, baseResolvedType, filter,
+				existingProperties, diagnostic, codeActions);
 
-		// Resolve methods defined in built-in Qute value resolvers
+		// Collect similar code action for methods defined in built-in Qute value
+		// resolvers
 		List<MethodValueResolver> resolvers = javaCache.getResolversFor(baseResolvedType, template.getProjectUri());
 		for (MethodValueResolver method : resolvers) {
 			if (method.isValidName()) {
-				availableValues.add(method.getMethodName());
+				doCodeActionsForSimilarValue(part, method.getMethodName(), template, existingProperties, diagnostic,
+						codeActions);
 			}
 		}
-
-		return availableValues;
 	}
 
-	private Collection<String> collectJavaTypeMethodNames(MethodPart part, Template template, String projectUri,
-	ResolvedJavaTypeInfo baseResolvedType, JavaTypeFilter filter, QuteNativeSettings nativeImageSettings) {
-		Collection<String> availableMethodNames = new HashSet<>();
-
-		List<String> javaMethodNames = baseResolvedType.getMethods().stream().map(x -> x.getMethodName())
-				.collect(Collectors.toList());
-		for (String methodName : javaMethodNames) {
-			availableMethodNames.add(methodName);
+	private void collectSimilarCodeActionsForJavaMethods(MethodPart part, Template template, String projectUri,
+			ResolvedJavaTypeInfo baseResolvedType, JavaTypeFilter filter, Set<String> existingProperties,
+			Diagnostic diagnostic, List<CodeAction> codeActions) {
+		// Java method similar code actions
+		for (JavaMethodInfo method : baseResolvedType.getMethods()) {
+			doCodeActionsForSimilarValue(part, method.getName(), template, existingProperties, diagnostic, codeActions);
 		}
 
+		// Java super method similar code actions
 		JavaTypeAccessibiltyRule javaTypeAccessibility = filter.getJavaTypeAccessibility(baseResolvedType,
 				template.getJavaTypesSupportedInNativeMode());
-
 		if (!isIgnoreSuperclasses(baseResolvedType, javaTypeAccessibility, filter)) {
 			List<String> extendedTypes = baseResolvedType.getExtendedTypes();
 			if (extendedTypes != null) {
 				for (String extendedType : extendedTypes) {
 					ResolvedJavaTypeInfo resolvedExtendedType = javaCache.resolveJavaType(extendedType, projectUri)
 							.getNow(null);
 					if (resolvedExtendedType != null) {
-						availableMethodNames.addAll(collectJavaTypeMethodNames(part, template, projectUri, resolvedExtendedType, filter, nativeImageSettings));
+						collectSimilarCodeActionsForJavaMethods(part, template, projectUri, resolvedExtendedType,

```

Comment:
If you have a cyclic inheritance chain eg. ClassB extends ClassA, ClassC extends ClassB, ClassA extends ClassC, then this will lead to infinite recursion when you open completion, and the result will never return.

It'd be nice to detect this or maybe put a limit on the number of times this can recurse.

Reply:
We could check that but I'm not sure that JDT allows that.

Could we see that in an another issues since there are a ot of code which does that. If the fix should be done,it should be done for several piece of code.

Reply:
I'll file this as a bug

Reply:
#725